src/streaming/zarr.stream.cpp

```diff
@@ -414,54 +600,409 @@ ZarrStream_s::create_store_()
 }
 
 void
-ZarrStream_s::write_multiscale_frames_(const uint8_t* data,
+ZarrStream_s::write_multiscale_frames_(const std::byte* data,
                                        size_t bytes_of_data)
 {
-    // TODO (aliddell): implement this
+    if (!multiscale_) {
+        return;
+    }
+
+    std::function<std::byte*(const std::byte*, size_t&, size_t&, size_t&)> scale;
+    std::function<void(void*, size_t, const void*, size_t)> average2;
+
+    switch (dtype_) {
+        case ZarrDataType_uint8:
+            scale = scale_image<uint8_t>;
+            average2 = average_two_frames<uint8_t>;
+            break;
+        case ZarrDataType_uint16:
+            scale = scale_image<uint16_t>;
+            average2 = average_two_frames<uint16_t>;
+            break;
+        case ZarrDataType_uint32:
+            scale = scale_image<uint32_t>;
+            average2 = average_two_frames<uint32_t>;
+            break;
+        case ZarrDataType_uint64:
+            scale = scale_image<uint64_t>;
+            average2 = average_two_frames<uint64_t>;
+            break;
+        case ZarrDataType_int8:
+            scale = scale_image<int8_t>;
+            average2 = average_two_frames<int8_t>;
+            break;
+        case ZarrDataType_int16:
+            scale = scale_image<int16_t>;
+            average2 = average_two_frames<int16_t>;
+            break;
+        case ZarrDataType_int32:
+            scale = scale_image<int32_t>;
+            average2 = average_two_frames<int32_t>;
+            break;
+        case ZarrDataType_int64:
+            scale = scale_image<int64_t>;
+            average2 = average_two_frames<int64_t>;
+            break;
+        case ZarrDataType_float32:
+            scale = scale_image<float>;
+            average2 = average_two_frames<float>;
+            break;
+        case ZarrDataType_float64:
+            scale = scale_image<double>;
+            average2 = average_two_frames<double>;
+            break;
+        default:
+            throw std::runtime_error("Invalid data type: " +
+                                     std::to_string(dtype_));
+    }
+
+    size_t frame_width = dimensions_->width_dim().array_size_px;
+    size_t frame_height = dimensions_->height_dim().array_size_px;
+
+    std::byte* dst;
+    for (auto i = 1; i < writers_.size(); ++i) {
+        dst = scale(data, bytes_of_data, frame_width, frame_height);
+
+        // bytes_of data is now downscaled
+        // frame_width and frame_height are now the new dimensions
+
+        if (scaled_frames_[i]) {
+            average2(dst, bytes_of_data, *scaled_frames_[i], bytes_of_data);
+            std::span frame_data{ reinterpret_cast<const std::byte*>(dst),
+                                  bytes_of_data };
+            EXPECT(writers_[i]->write_frame(frame_data),
+                   "Failed to write frame to writer %zu",
+                   i);
+
+            // clean up this LOD
+            delete[] *scaled_frames_[i];
```

Comment:
Have you considered managing resource lifetimes with smart pointers? I think that modern C++ code should avoid naked new/delete operations, except when dealing with legacy code compatibility or API integration.